src/main/java/org/mockito/internal/creation/bytebuddy/TypeCachingBytecodeGenerator.java

```diff
@@ -36,7 +39,7 @@ public TypeCachingBytecodeGenerator(BytecodeGenerator bytecodeGenerator, boolean
                         public Class<?> call() throws Exception {
                             return bytecodeGenerator.mockClass(params);
                         }
-                    }, classLoader == null ? BOOTSTRAP_LOCK : classLoader);
+                    }, singletonLock || classLoader == null ? BOOTSTRAP_LOCK : classLoader);
```

Comment:
@raphw, can we lock always using BOOTSTRAP_LOCK?

In JVM that runs tests, we have one instance of BOOTSTRAP_LOCK and (typically) one instance of the classloader. If using classloader is a performance tweak, do we have an evidence how much value it adds? Also, the lock is only used for "new" classes (not yet in cache). I am not convinced using "classloader" as monitor adds value.

It is a best practice to always use internal objects as synchronization monitors (private final Object). This way, nobody else will use that object for monitoring and we're safe from deadlocks.

I'm not a fan of adding a system property to control this. It does not help with user experience because the user is impacted by a hanging build. After some googling, the user might be able to find an answer or he needs to reach out to us to learn about the property. But then what - he needs to remember to use the property in codebases that are exposed to this problem? Also, a system property is yet another method to configure Mockito - let's limit the various ways Mockito can be configured.

Hope you don't mind me complaining :) I'm super grateful you debugged this issue and offered a solution! For Mockito, it's best if we keep things simple, add new configuration options only when we really need to, and strive for great user experience.